File: frontend/src/services/api/recommendationService.ts
/**
 * 추천·이력 서비스 API 함수
 * recommendation-service 실제 컨트롤러 엔드포인트와 정확히 일치
 *
 * 응답 구조: ApiResponse<T> { success, data, error, timestamp }
 * 호출부에서 res.data.data 로 실제 페이로드에 접근한다.
 */
import { recommendationApiClient } from './instances'
import type { ApiResponse } from '@/types/api'
import type {
  TodayRecommendationsResponse,
  RecommendationReasonResponse,
  AcceptRecommendationRequest,
  AcceptRecommendationResponse,
  RejectRecommendationRequest,
  RejectRecommendationResponse,
  RefreshRecommendationsRequest,
  CreateMealRequest,
  UpdateMealRequest,
  MealResponse,
  FeedbackRequest,
  FeedbackResponse,
  MealHistoryResponse,
  InsightsResponse,
} from '@/types/recommendation'

export const recommendationService = {
  /** GET /api/v1/recommendations/today — 오늘의 추천 3개 조회 */
  getTodayRecommendations: (params: { latitude: number; longitude: number; requestedAt?: string }) =>
    recommendationApiClient.get<ApiResponse<TodayRecommendationsResponse>>('/recommendations/today', { params }),

  /** GET /api/v1/recommendations/{recommendationId}/reason — 추천 이유 상세 확인 */
  getRecommendationReason: (recommendationId: string) =>
    recommendationApiClient.get<ApiResponse<RecommendationReasonResponse>>(
      `/recommendations/${recommendationId}/reason`,
    ),

  /** POST /api/v1/recommendations/{recommendationId}/accept — 추천 수락 */
  acceptRecommendation: (recommendationId: string, data: AcceptRecommendationRequest) =>
    recommendationApiClient.post<ApiResponse<AcceptRecommendationResponse>>(
      `/recommendations/${recommendationId}/accept`,
      data,
    ),

  /** POST /api/v1/recommendations/{recommendationId}/reject — 추천 거절 */
  rejectRecommendation: (recommendationId: string, data: RejectRecommendationRequest) =>
    recommendationApiClient.post<ApiResponse<RejectRecommendationResponse>>(
      `/recommendations/${recommendationId}/reject`,
      data,
    ),

  /** POST /api/v1/recommendations/refresh — 추천 전체 새로고침 */
  refreshRecommendations: (data: RefreshRecommendationsRequest) =>
    recommendationApiClient.post<ApiResponse<TodayRecommendationsResponse>>('/recommendations/refresh', data),

  /** POST /api/v1/meals — 식사 완료 원탭 기록 */
  createMeal: (data: CreateMealRequest) =>
    recommendationApiClient.post<ApiResponse<MealResponse>>('/meals', data),

  /** PUT /api/v1/meals/{mealId} — 식사 기록 수정 */
  updateMeal: (mealId: string, data: UpdateMealRequest) =>
    recommendationApiClient.put<ApiResponse<MealResponse>>(`/meals/${mealId}`, data),

  /** DELETE /api/v1/meals/{mealId} — 식사 기록 취소 (30초 이내) */
  deleteMeal: (mealId: string) =>
    recommendationApiClient.delete<ApiResponse<{ message: string }>>(`/meals/${mealId}`),

  /** POST /api/v1/meals/{mealId}/feedback — 식사 피드백 제출 */
  submitFeedback: (mealId: string, data: FeedbackRequest) =>
    recommendationApiClient.post<ApiResponse<FeedbackResponse>>(`/meals/${mealId}/feedback`, data),

  /** GET /api/v1/history/timeline — 식사 이력 타임라인 조회 */
  getHistoryTimeline: (params?: { startDate?: string; endDate?: string }) =>
    recommendationApiClient.get<ApiResponse<MealHistoryResponse>>('/history/timeline', { params }),

  /** GET /api/v1/insights — 취향 인사이트 리포트 조회 */
  getInsights: () =>
    recommendationApiClient.get<ApiResponse<InsightsResponse>>('/insights'),
}

File: recommendation-service/src/main/java/com/unicorn/lunchpick/recommendation/service/RecommendationService.java
package com.unicorn.lunchpick.recommendation.service;

import com.unicorn.lunchpick.recommendation.dto.request.AcceptRecommendationRequest;
import com.unicorn.lunchpick.recommendation.dto.request.RefreshRecommendationsRequest;
import com.unicorn.lunchpick.recommendation.dto.request.RejectRecommendationRequest;
import com.unicorn.lunchpick.recommendation.dto.response.AcceptRecommendationResponse;
import com.unicorn.lunchpick.recommendation.dto.response.RecommendationReasonResponse;
import com.unicorn.lunchpick.recommendation.dto.response.RejectRecommendationResponse;
import com.unicorn.lunchpick.recommendation.dto.response.TodayRecommendationsResponse;

/**
 * 추천 서비스 인터페이스
 *
 * <p>오늘의 추천 조회, 추천 이유 조회, 수락/거절/새로고침을 담당합니다.</p>
 *
 * @author lunchpick-team
 * @version 1.0
 * @since 2026-02-26
 */
public interface RecommendationService {

    /**
     * 오늘의 추천 3개 조회
     *
     * <p>캐시 히트 시 Redis에서 반환, 미스 시 AI Pipeline 호출(Resilience4j CB 보호).
     * AI Pipeline 장애 시 규칙 기반 폴백 추천을 반환합니다.</p>
     *
     * @param memberId  회원 식별자
     * @param latitude  위도
     * @param longitude 경도
     * @return 오늘의 추천 응답
     */
    TodayRecommendationsResponse getTodayRecommendations(String memberId, double latitude, double longitude);

    /**
     * 추천 이유 상세 조회
     *
     * @param memberId         회원 식별자
     * @param recommendationId 추천 ID
     * @return 추천 이유 응답
     */
    RecommendationReasonResponse getRecommendationReason(String memberId, String recommendationId);

    /**
     * 추천 수락
     *
     * @param memberId         회원 식별자
     * @param recommendationId 추천 ID
     * @param request          수락 요청
     * @return 수락 응답
     */
    AcceptRecommendationResponse acceptRecommendation(String memberId, String recommendationId,
                                                       AcceptRecommendationRequest request);

    /**
     * 추천 거절 및 대체 추천 반환
     *
     * @param memberId         회원 식별자
     * @param recommendationId 추천 ID
     * @param request          거절 요청
     * @return 거절 응답 (대체 추천 포함 또는 없음)
     */
    RejectRecommendationResponse rejectRecommendation(String memberId, String recommendationId,
                                                       RejectRecommendationRequest request);

    /**
     * 추천 전체 새로고침 (3개 모두 거절 시)
     *
     * @param memberId 회원 식별자
     * @param request  새로고침 요청 (거절 ID 목록 + 위치)
     * @return 새로운 추천 3개
     */
    TodayRecommendationsResponse refreshRecommendations(String memberId,
                                                         RefreshRecommendationsRequest request);
}
